src/UI/MainWindow.vala

```diff
@@ -202,7 +203,72 @@ public class Spreadsheet.UI.MainWindow : ApplicationWindow {
                 app_stack.set_visible_child_name ("app");
             }
         });
-        return welcome;
+
+        var box = new Box (Orientation.HORIZONTAL, 0);
+
+        // This is so that the background becomes white - there may be a better way
+        box.get_style_context ().add_class (Gtk.STYLE_CLASS_VIEW);
+
+        var recent_files_box = recent_files ();
+
+        box.pack_start (welcome);
+        box.pack_start (new Separator (Orientation.VERTICAL));
+        box.pack_start (recent_files_box);
+        return box;
+    }
+
+    private Widget recent_files () {
+        var title = new Label (_("Recent files"));
+
+        title.get_style_context ().add_class (Granite.STYLE_CLASS_H2_LABEL);
+
+        update_listview ();
+
+        var scrolledWindow = new ScrolledWindow (null, null);
+        scrolledWindow.hscrollbar_policy = PolicyType.NEVER;
+
+        var recent_files_box = new Box (Orientation.VERTICAL, 0);
+        recent_files_box.pack_start (title);
+        recent_files_box.pack_start (list_view);
+        
+        scrolledWindow.add (recent_files_box);
+        return scrolledWindow;
+    }
+
+    private void update_listview () {
+        foreach (var item in list_view.get_children ()) {
+            item.destroy ();
+        }
+
+        var recent_files = Spreadsheet.App.settings.get_strv ("recent-files");
+
+        foreach (var file_name in recent_files) {
+        	var file = File.new_for_path (file_name);
+        	if (file.query_exists ()) {
+        	    var basename = file.get_basename ();
+            	var path = file.get_path ();
+                // IconSize.DIALOG because it's 48px, just like WelcomeButton needs
+                var spreadsheet_icon = new Image.from_icon_name ("x-office-spreadsheet",
+                    IconSize.DIALOG);
+                var list_item = new WelcomeButton (spreadsheet_icon, basename, path);
+                list_item.clicked.connect (() => {
+                    try {
+                        this.file = new CSVParser.from_file (path).parse ();
+                        init_header ();
+                        show_all ();
+                        app_stack.set_visible_child_name ("app");
+                    } catch (ParserError err) {
+                        debug ("Error: " + err.message);
+                    }
+                });
+                list_view.add (list_item);
+        	} else {
+        	    /* In case the file doesn't exist, display a list item, but
+        	       mark the file as missing? */
```

Comment:
What about showing an error dialog like this, when users click an unavailable file in the list?

![Screenshot from 2020-02-01 20-56-15](https://user-images.githubusercontent.com/26003928/73591710-63e8b600-4535-11ea-8233-9f1a2aac71d1.png)

Maybe something like this:

<details>

```diff
         foreach (var file_name in recent_files) {
                var file = File.new_for_path (file_name);
-               if (file.query_exists ()) {
-                   var basename = file.get_basename ();
-               var path = file.get_path ();
-                // IconSize.DIALOG because it's 48px, just like WelcomeButton needs
-                var spreadsheet_icon = new Image.from_icon_name ("x-office-spreadsheet",
-                    IconSize.DIALOG);
-                var list_item = new WelcomeButton (spreadsheet_icon, basename, path);
-                list_item.clicked.connect (() => {
-                    try {
-                        this.file = new CSVParser.from_file (path).parse ();
-                        init_header ();
-                        show_all ();
-                        app_stack.set_visible_child_name ("app");
-                    } catch (ParserError err) {
-                        debug ("Error: " + err.message);
-                    }
-                });
-                list_view.add (list_item);
-               } else {
-                   /* In case the file doesn't exist, display a list item, but
-                      mark the file as missing? */
-               }
+            var basename = file.get_basename ();
+            var path = file.get_path ();
+            // IconSize.DIALOG because it's 48px, just like WelcomeButton needs
+            var spreadsheet_icon = new Image.from_icon_name ("x-office-spreadsheet",
+                IconSize.DIALOG);
+            var list_item = new WelcomeButton (spreadsheet_icon, basename, path);
+            list_item.clicked.connect (() => {
+                if (!file.query_exists ()) {
+                    var file_not_found_dialog = new Granite.MessageDialog.with_image_from_icon_name (
+                        _("Failed to open file"),
+                        _("The file could not found."),
+                        "dialog-error", Gtk.ButtonsType.CLOSE);
+                    file_not_found_dialog.transient_for = this;
+                    file_not_found_dialog.run ();
+                    file_not_found_dialog.destroy ();
+
+                    return;
+                }
+
+                try {
+                    this.file = new CSVParser.from_file (path).parse ();
+                    init_header ();
+                    show_all ();
+                    app_stack.set_visible_child_name ("app");
+                } catch (ParserError err) {
+                    debug ("Error: " + err.message);
+                }
+            });
+            list_view.add (list_item);
         }
 
         list_view.show_all ();
```

</details>

Reply:
Commit 67d8665 resolved this by removing the item from the list if the file is no longer exists.